sklearn/utils/validation.py

```diff
@@ -476,6 +476,13 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
             # To ensure that array flags are maintained
             array = np.array(array, dtype=dtype, order=order, copy=copy)
 
+        # in the future np.flexible dtypes will be handled like object dtypes
+        if dtype_numeric and np.issubdtype(array.dtype, np.flexible):
+            warnings.warn(
+                "In the future, array with dtype {} will be handled as object,"
```

Comment:
Plese follow the conventions of the message given in http://scikit-learn.org/dev/developers/contributing.html#deprecation.

Unless I miss something you should say explicitly in the message that it will cause an error (rather than handled as objects which probably does not mean much to the end user).

Reply:
Here's an example where I think it would make sense to handle in the same way as an object array rather than raise an error:
let `X_str_num = [['1', '2'], ['3', '4']]` then `check_array(np.array(X_str_num, 'O'), 'numeric')` returns an array with dtype float64 but `check_array(X_str_num, 'numeric')` and `check_array(np.array(X_str_num, 'U1'), 'numeric')` return an array with dtype U1. 

This behavior for object arrays is expected base on the docstring, so we would just have to expand that description from object to object + flexible at the end of the deprecation cycle if we do want to treat flexible arrays the same as objects (e.g., attempt to cast to numeric).
> If "numeric", dtype is preserved unless array.dtype is object.

Reply:
To align the deprecation warning message with the guidelines, I've added the version the change in behavior is to be expected to the warning in commit 4bcfc92. I'm not sure if we need to indicate the current version, since no changes have actually been made this version. For this reason I also haven't added a deprecation note to the docstring, but please let me know if you'd suggest otherwise.